core/src/main/java/io/grpc/internal/SpiffeUtil.java

```diff
@@ -96,6 +126,143 @@ private static void validatePathSegment(String pathSegment) {
+    Object jsonObject = JsonParser.parseNoDuplicates(json);
+    if (!(jsonObject instanceof Map)) {
+      throw new IllegalArgumentException(
+          "SPIFFE Trust Bundle should be a JSON object. Found: "
+              + (jsonObject == null ? null : jsonObject.getClass()));
+    }
+    @SuppressWarnings("unchecked")
+    Map<String, ?> root = (Map<String, ?>)jsonObject;
+    Map<String, ?> trustDomainsNode = JsonUtil.getObject(root, "trust_domains");
+    checkArgument(checkNotNull(trustDomainsNode,
+            "Mandatory trust_domains element is missing").size() > 0,
+        "Mandatory trust_domains element is missing");
+    return trustDomainsNode;
+  }
+
+  private static boolean checkJwkEntry(Map<String, ?> jwkNode, String trustDomainName) {
+    String kty = JsonUtil.getString(jwkNode, "kty");
+    if (kty == null || !kty.equals(KTY_PARAMETER_VALUE)) {
+      log.log(Level.SEVERE, String.format("'kty' parameter must be '%s' but '%s' found. "
+              + "Skipping certificate loading for trust domain '%s'.", KTY_PARAMETER_VALUE, kty,
+          trustDomainName));
+      return false;
+    }
+    String kid = JsonUtil.getString(jwkNode, "kid");
+    if (kid != null && !kid.equals("")) {
+      log.log(Level.SEVERE, String.format("'kid' parameter must not be set but value '%s' "
+              + "found. Skipping certificate loading for trust domain '%s'.", kid,
+          trustDomainName));
+      return false;
+    }
+    String use = JsonUtil.getString(jwkNode, "use");
+    if (use == null || !use.equals(USE_PARAMETER_VALUE)) {
+      log.log(Level.SEVERE, String.format("'use' parameter must be '%s' but '%s' found. "
+              + "Skipping certificate loading for trust domain '%s'.", USE_PARAMETER_VALUE, use,
+          trustDomainName));
+      return false;
+    }
+    return true;
+  }
+
+  private static List<X509Certificate> extractCert(List<Map<String, ?>> keysNode,
+      String trustDomainName) {
+    List<X509Certificate> result = new ArrayList<>();
+    for (Map<String, ?> keyNode : keysNode) {
+      if (!checkJwkEntry(keyNode, trustDomainName)) {
+        break;
+      }
+      List<String> rawCerts = JsonUtil.getListOfStrings(keyNode, "x5c");
+      if (rawCerts == null) {
+        break;
+      }
+      if (rawCerts.size() != 1) {
+        log.log(Level.SEVERE, String.format("Exactly 1 certificate is expected, but %s found for "
+            + "domain %s.", rawCerts.size(), trustDomainName));
+        break;
+      }
+      InputStream stream = new ByteArrayInputStream((CERTIFICATE_PREFIX + System.lineSeparator()
+          + rawCerts.get(0) + System.lineSeparator() + CERTIFICATE_SUFFIX)
+          .getBytes(StandardCharsets.UTF_8));
+      try {
+        Collection<? extends Certificate> certs = CertificateFactory.getInstance("X509")
+            .generateCertificates(stream);
+        result.add(certs.toArray(new X509Certificate[0])[0]);
```

Comment:
Should we check that `certs.size()` is exactly 1?

Reply:
I believe we do it here - https://github.com/grpc/grpc-java/blob/821810f51b029ed4cef30eb421e3b4bd01c913ab/core/src/main/java/io/grpc/internal/SpiffeUtil.java#L246
IIUC since we manually add BEGIN and END CERT, then it's not possible to have a string (single element of `rawCerts` with multiple certs in it.

Reply:
But could it be empty if `generateCertificates` fails somehow?

Reply:
Yeah that's a tricky question. I couldn't create a test to make it return empty collection without failing (knowing the previous check of `rawCerts`), **but** the Javadoc for this method doesn't really guarantee that - https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/security/cert/CertificateFactory.html#generateCertificates(java.io.InputStream)
Keeping in mind that other JDKs might implement it slightly different, I think it's worth to add such check (I used `checkArgument` in that case, maybe some `assert` is more appropriate).